cdc_iterator.go

```diff
@@ -0,0 +1,303 @@
+	c, err := common.NewCanal(config.mysqlConfig, config.tables)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create canal: %w", err)
+	}
+
+	sdk.Logger(ctx).Info().Msg("created canal")
+
+	pos, err := c.GetMasterPos()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get master pos when creating cdc iterator: %w", err)
+	}
+
+	rowsEventsC := make(chan rowEvent)
+	canalDoneC := make(chan struct{})
+	eventHandler := newCdcEventHandler(pos.Name, canalDoneC, rowsEventsC)
+
+	iterator := &cdcIterator{
+		canal:        c,
+		rowsEventsC:  rowsEventsC,
+		canalRunErrC: make(chan error),
+		canalDoneC:   canalDoneC,
+		config:       config,
+	}
+
+	startPosition, err := iterator.getStartPosition(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get start position: %w", err)
+	}
+
+	go func() {
+		iterator.canal.SetEventHandler(eventHandler)
+		iterator.canalRunErrC <- iterator.canal.RunFrom(startPosition)
+	}()
+
+	return iterator, nil
+}
+
+func (c *cdcIterator) getStartPosition(config cdcIteratorConfig) (common.CdcPosition, error) {
+	var cdcPosition common.CdcPosition
+	if config.position != nil {
+		pos, err := common.ParseSDKPosition(config.position)
+		if err != nil {
+			return cdcPosition, fmt.Errorf("failed to parse position: %w", err)
+		}
+		if pos.Kind == common.PositionTypeSnapshot {
+			return cdcPosition, fmt.Errorf("invalid position type: %s", pos.Kind)
+		}
+
+		return *pos.CdcPosition, nil
+	}
+
+	masterPos, err := c.canal.GetMasterPos()
+	if err != nil {
+		return cdcPosition, fmt.Errorf("failed to get master position: %w", err)
+	}
+
+	return common.CdcPosition{
+		Name: masterPos.Name,
+		Pos:  masterPos.Pos,
+	}, nil
+}
+
+func (c *cdcIterator) Ack(context.Context, sdk.Position) error {
+	return nil
+}
+
+func (c *cdcIterator) Next(ctx context.Context) (rec sdk.Record, err error) {
+	select {
+	case <-ctx.Done():
+		//nolint:wrapcheck // no need to wrap canceled error
+		return rec, ctx.Err()
+	case <-c.canalDoneC:
+		return rec, fmt.Errorf("canal is closed")
+	case data := <-c.rowsEventsC:
+		rec, err := c.buildRecord(data)
+		if err != nil {
+			return rec, fmt.Errorf("failed to build record: %w", err)
+		}
+
+		return rec, nil
+	}
+}
+
+func (c *cdcIterator) Teardown(ctx context.Context) error {
+	close(c.canalDoneC)
+
+	c.canal.Close()
+	select {
+	case <-ctx.Done():
+		//nolint:wrapcheck // no need to wrap canceled error
+		return ctx.Err()
+	case err := <-c.canalRunErrC:
+		if err != nil {
+			return fmt.Errorf("failed to stop canal: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func buildPayload(columns []schema.TableColumn, rows []any) sdk.StructuredData {
+	payload := sdk.StructuredData{}
+	for i, col := range columns {
+		payload[col.Name] = common.FormatValue(rows[i])
+	}
+	return payload
+}
+
+func (c *cdcIterator) buildRecord(e rowEvent) (sdk.Record, error) {
+	pos := common.CdcPosition{
+		Name: e.binlogName,
+		Pos:  e.Header.LogPos,
+	}
+
+	metadata := sdk.Metadata{}
```

Comment:
Let's also add the [server ID](https://github.com/conduitio-labs/conduit-connector-mysql/issues/36) to the metadata.